overlord/snapshotstate/backend/backend.go

```diff
@@ -0,0 +1,220 @@
+				err = nil
+				continue
+			}
+			err = f(rsh)
+			if err != nil {
+				break
+			}
+			err = rsh.Close()
+			if err != nil {
+				break
+			}
+		}
+	}
+
+	if err == io.EOF {
+		err = nil
+	}
+
+	return err
+}
+
+// List valid snapshots
+//
+// The snapshots are closed before returning.
+func List(ctx context.Context, snapshotID uint64, snapNames []string) ([]client.SnapshotGroup, error) {
+	shotsmap := map[uint64][]*client.Snapshot{}
+	err := Iter(ctx, func(sh *Reader) error {
+		if snapshotID == 0 || sh.ID == snapshotID {
+			if len(snapNames) == 0 || strutil.ListContains(snapNames, sh.Snap) {
+				shotsmap[sh.ID] = append(shotsmap[sh.ID], &sh.Snapshot)
+			}
+		}
+		return nil
+	})
+
+	groups := make([]client.SnapshotGroup, 0, len(shotsmap))
+	for id, shots := range shotsmap {
+		sort.Sort(bySnap(shots))
+		groups = append(groups, client.SnapshotGroup{ID: id, Snapshots: shots})
+	}
+
+	sort.Sort(byID(groups))
+
+	return groups, err
+}
+
+var isValidVersion = regexp.MustCompile("^[a-zA-Z0-9.+~-]{1,32}$").MatchString
+
+// Filename of the given client.Snapshot in this backend.
+func Filename(sh *client.Snapshot) string {
+	var basename string
+	if snap.ValidateName(sh.Snap) == nil {
+		if isValidVersion(sh.Version) {
+			basename = fmt.Sprintf("%s_%s_%d.zip", sh.Snap, sh.Version, sh.ID)
+		} else {
+			// AFAIK snapcraft would fail these, but we're more generous
+			basename = fmt.Sprintf("%s_rev%s_%d.zip", sh.Snap, sh.Revision, sh.ID)
+		}
+	} else {
+		// just give up
+		basename = fmt.Sprintf("%d.zip", sh.ID)
+	}
+	return filepath.Join(dirs.SnapshotDir, basename)
+}
+
+// Save a snapshot
+func Save(ctx context.Context, id uint64, si *snap.Info, cfg *json.RawMessage, homes []string) (sh *client.Snapshot, e error) {
+	if err := os.MkdirAll(dirs.SnapshotDir, 0700); err != nil {
+		return nil, err
+	}
+
+	sh = &client.Snapshot{
+		ID:       id,
+		Snap:     si.Name(),
+		Revision: si.Revision,
+		Version:  si.Version,
+		Time:     time.Now(),
+		Hashsums: make(map[string]string),
+		Size:     0,
+		Config:   cfg,
+	}
+
+	aw, err := osutil.NewAtomicFile(Filename(sh), 0600, 0, osutil.NoChown, osutil.NoChown)
+	if err != nil {
+		return nil, err
+	}
+	// if things worked, we'll commit (and Cancel becomes a NOP)
+	defer aw.Cancel()
+
+	w := zip.NewWriter(aw)
+	hasher := crypto.SHA3_384.New()
+	if err := addToZip(ctx, sh, w, archiveName, si.DataDir(), hasher); err != nil {
+		return nil, err
+	}
+	hasher.Reset()
+
+	for _, home := range homes {
+		if err := addToZip(ctx, sh, w, userArchiveName(home), si.UserDataDir(home), hasher); err != nil {
+			return nil, err
+		}
+		hasher.Reset()
+	}
+
+	metaWriter, err := w.Create(metadataName)
+	if err != nil {
+		return nil, err
+	}
+
+	enc := json.NewEncoder(io.MultiWriter(metaWriter, hasher))
+	if err := enc.Encode(sh); err != nil {
+		return nil, err
+	}
+
+	hashWriter, err := w.Create(metaHashName)
+	if err != nil {
+		return nil, err
+	}
+	fmt.Fprintf(hashWriter, "%x\n", hasher.Sum(nil))
+	w.Close()
```

Comment:
Should that be `defer w.Close()` earlier?

Reply:
zips are weird. This `w.Close()` (whose return value I should check, ouch!) does not close the file (that's `aw`), just writes out the 'central directory' of the zip.